File: crons/src/main/java/com/raxdenstudios/cron/utils/CronUtils.java
package com.raxdenstudios.cron.utils;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;

import com.raxdenstudios.commons.util.ConvertUtils;
import com.raxdenstudios.commons.util.Utils;
import com.raxdenstudios.cron.model.Cron;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

/**
 * Created by Angel on 07/02/2018.
 */

public class CronUtils {

    public static void setAlarmManager(Context context, Cron cron) {
        if (cron != null && cron.getTriggerAtTime() > 0) {
            AlarmManager manager = createAlarmManager(context);
            PendingIntent pendingIntent = initPendingIntent(context, cron);
            if (manager != null && pendingIntent != null) {
                // Schedule the cron!
                manager.cancel(pendingIntent);
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                    manager.setExact(cron.getType(), cron.getTriggerAtTime(), pendingIntent);
                } else {
                    manager.set(cron.getType(), cron.getTriggerAtTime(), pendingIntent);
                }
            }
        }
    }

    public static void cancelAlarmManager(Context context, Cron cron) {
        if (cron != null) {
            AlarmManager manager = createAlarmManager(context);
            PendingIntent pendingIntent = initPendingIntent(context, cron);
            if (pendingIntent != null) {
                // Cancel the cron!
                manager.cancel(pendingIntent);
            }
        }
    }

    private static PendingIntent initPendingIntent(Context context, Cron cron) {
        Intent intent = new Intent();
        intent.setAction(Utils.getPackageName(context) + ".CRON");
        intent.putExtra(Cron.class.getSimpleName(), cron.getId());
        return PendingIntent.getService(context, ConvertUtils.longToInt(cron.getId()), intent, 0);
    }

    private static AlarmManager createAlarmManager(Context context) {
        return (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
    }

    public static String currentDateTime() {
        return new SimpleDateFormat("dd/MM/yy HH:mm:ss", Locale.getDefault()).format(new Date());
    }

    public static String triggerAtTime(Cron cron) {
        return new SimpleDateFormat("dd/MM/yy HH:mm:ss", Locale.getDefault()).format(new Date(cron.getTriggerAtTime()));
    }

    public static String intervalInSeconds(Cron cron) {
        return ((float) (cron.getInterval() / (1000))) + " seconds";
    }

    public static String nextLaunchInSeconds(Cron cron) {
        return ((float) ((cron.getTriggerAtTime() - Calendar.getInstance().getTimeInMillis()) / (1000))) + " seconds";
    }

}

File: crons/src/main/java/com/raxdenstudios/cron/utils/CronUtils.kt
package com.raxdenstudios.cron.utils

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import android.util.Log
import com.raxdenstudios.cron.CronHandler
import com.raxdenstudios.cron.CronManager
import com.raxdenstudios.cron.model.Cron
import java.text.SimpleDateFormat
import java.util.*

object CronUtils {

    private val TAG = CronUtils::class.java.simpleName

    internal var isForeground: Boolean = false

    fun setAlarmManager(context: Context, cron: Cron) {
        cron.takeIf { it.triggerAtTime > 0 }?.apply {
            createAlarmManager(context).apply {
                val pendingIntent = initPendingIntent(context, cron)
                cancel(pendingIntent)
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                    setExact(cron.type, cron.triggerAtTime, pendingIntent)
                } else {
                    set(type, cron.triggerAtTime, pendingIntent)
                }
            }
        }
    }

    fun cancelAlarmManager(context: Context, cron: Cron) {
        createAlarmManager(context).apply {
            val pendingIntent = initPendingIntent(context, cron)
            cancel(pendingIntent)
        }
    }

    fun currentDateTime() = SimpleDateFormat("dd/MM/yy HH:mm:ss", Locale.getDefault()).format(Date())

    fun triggerAtTime(cron: Cron) = SimpleDateFormat("dd/MM/yy HH:mm:ss", Locale.getDefault()).format(Date(cron.triggerAtTime))

    fun intervalInSeconds(cron: Cron) = StringBuffer((cron.interval / 1000).toFloat().toString()).append(" seconds")

    fun nextLaunchInSeconds(cron: Cron) = StringBuffer(((cron.triggerAtTime - Calendar.getInstance().timeInMillis) / 1000).toFloat().toString()).append(" seconds")

    private fun createAlarmManager(context: Context) = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager

    private fun initPendingIntent(context: Context, cron: Cron): PendingIntent {
        val intent = Intent().apply {
            action = getPackageName(context) + ".CRON"
            putExtra(Cron::class.java.simpleName, cron.id)
        }
        return if (!isForeground && Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            Log.d(TAG, "PendingIntent.getForegroundService....")
            PendingIntent.getForegroundService(context, cron.id.toInt(), intent, 0)
        } else {
            Log.d(TAG, "PendingIntent.getService....")
            return PendingIntent.getService(context, cron.id.toInt(), intent, 0)
        }
    }

    private fun getPackageName(context: Context) = context.packageManager.getPackageInfo(context.packageName, 0).packageName

}